src/Control/Monad/Infer.hs

```diff
@@ -53,9 +55,16 @@ instance Semigroup Env where
   Env a b <> Env a' b' = Env (a <> a') (b <> b')
 
 data Constraint p
-  = ConUnify (Expr p) (Type p) (Type p)
+  = ConUnify SomeReason (Type p) (Type p)
+  | ConSubsume SomeReason (Type p) (Type p)
 deriving instance (Show (Ann p), Show (Var p), Show (Expr p), Show (Type p))
   => Show (Constraint p)
+deriving instance (Eq (Ann p), Eq (Var p), Eq (Expr p), Eq (Type p))
+  => Eq (Constraint p)
+
+instance Eq (Constraint p) => Ord (Constraint p) where
```

Comment:
Are you sure this implementation is acceptable?

Reply:
It's not "acceptable", but it's kinda needed: We have to solve unification constraints before subsumption ones otherwise stuff will go bad. 

Stuff is still currently going bad, but it'll go badder.